File: PE-common/jvm/src/main/java/nl/adaptivity/util/JAXBAdapter.java
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.util.xml;

import nl.adaptivity.xml.XmlSerializable;
import nl.adaptivity.xml.XmlStreaming;
import org.jetbrains.annotations.NotNull;
import org.w3c.dom.Document;
import org.w3c.dom.DocumentFragment;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

import javax.xml.bind.annotation.adapters.XmlAdapter;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.stream.XMLOutputFactory;
import javax.xml.stream.XMLStreamWriter;
import javax.xml.transform.dom.DOMResult;


/**
 * Created by pdvrieze on 11/04/16.
 */
public class JAXBAdapter extends XmlAdapter<SimpleAdapter, XmlSerializable> {

  @Override
  public XmlSerializable unmarshal(final SimpleAdapter v) throws Exception {
    throw new UnsupportedOperationException();
  }

  @NotNull
  @Override
  public SimpleAdapter marshal(@NotNull final XmlSerializable v) throws Exception {


    final DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
    dbf.setNamespaceAware(true);
    final Document         document = dbf.newDocumentBuilder().newDocument();
    final DocumentFragment content  = document.createDocumentFragment();
    final XMLOutputFactory xof      = XMLOutputFactory.newFactory();
    final XMLStreamWriter  out      = xof.createXMLStreamWriter(new DOMResult(content));

    v.serialize(XmlStreaming.newWriter(new DOMResult(content)));
    final int childCount = content.getChildNodes().getLength();
    if (childCount == 0) {
      return new SimpleAdapter();
    } else if (childCount == 1) {
      final SimpleAdapter result = new SimpleAdapter();
      final Node          child  = content.getFirstChild();
      if (child instanceof Element) {
        result.setAttributes(child.getAttributes());
        for (Node child2 = child.getFirstChild(); child2 != null; child2 = child2.getNextSibling()) {
          result.children.add(child2);
        }
      } else {
        result.children.add(child);
      }
      return result;
    } else { // More than one child
      final SimpleAdapter result = new SimpleAdapter();
      for (Node child = content.getFirstChild(); child != null; child = child.getNextSibling()) {
        result.children.add(child);
      }
      return result;
    }
  }
}

File: PE-common/src/jvmMain/kotlin/nl/adaptivity/util/JAXBAdapter.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.util

import nl.adaptivity.util.xml.SimpleAdapter
import nl.adaptivity.xmlutil.XmlSerializable
import nl.adaptivity.xmlutil.XmlStreaming
import org.w3c.dom.Document
import org.w3c.dom.DocumentFragment
import org.w3c.dom.Element
import org.w3c.dom.Node

import javax.xml.bind.annotation.adapters.XmlAdapter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.stream.XMLOutputFactory
import javax.xml.stream.XMLStreamWriter
import javax.xml.transform.dom.DOMResult


/**
 * Created by pdvrieze on 11/04/16.
 */
open class JAXBAdapter : XmlAdapter<SimpleAdapter, XmlSerializable>() {

    @Throws(Exception::class)
    override fun unmarshal(v: SimpleAdapter): XmlSerializable {
        throw UnsupportedOperationException()
    }

    @Throws(Exception::class)
    override fun marshal(v: XmlSerializable): SimpleAdapter {


        val dbf = DocumentBuilderFactory.newInstance()
        dbf.isNamespaceAware = true
        val document = dbf.newDocumentBuilder().newDocument()
        val content = document.createDocumentFragment()
        val xof = XMLOutputFactory.newFactory()
        val out = xof.createXMLStreamWriter(DOMResult(content))

        v.serialize(XmlStreaming.newWriter(DOMResult(content)))
        val childCount = content.childNodes.length
        if (childCount == 0) {
            return SimpleAdapter()
        } else if (childCount == 1) {
            val result = SimpleAdapter()
            val child = content.firstChild
            if (child is Element) {
                result.setAttributes(child.getAttributes())
                var child2: Node? = child.getFirstChild()
                while (child2 != null) {
                    result.children.add(child2)
                    child2 = child2.nextSibling
                }
            } else {
                result.children.add(child)
            }
            return result
        } else { // More than one child
            val result = SimpleAdapter()
            var child: Node? = content.firstChild
            while (child != null) {
                result.children.add(child)
                child = child.nextSibling
            }
            return result
        }
    }
}
